directorywatcher/directorywatcher.go

```diff
@@ -0,0 +1,288 @@
+
+// RemoveFile is a type of function that should be ran to handle removing
+// data after a file has been removed from the watcher
+type RemoveFile func(d interface{}, path string) (data interface{}, err error)
+
+// Result is the return type of New. Use Get function to get the actual data.
+type Result struct {
+	data atomic.Value
+
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+// Config defines the config to be used in New function.
+//
+// Can be deserialized from YAML.
+type Config struct {
+	// The path to the file to be watched, required.
+	Path string `yaml:"path"`
+
+	// The parser to parse the data load, required.
+	Parser filewatcher.Parser
+
+	Adder AddFile
+
+	Remover RemoveFile
+
+	// Optional. When non-nil, it will be used to log errors,
+	// either returned by parser or by the underlying file system watcher.
+	// Please note that this does not include errors returned by the first parser
+	// call, which will be returned directly.
+	Logger log.Wrapper `yaml:"logger"`
+
+	// Optional. When <=0 DefaultMaxFileSize will be used instead.
+	//
+	// This is the soft limit,
+	// we will also auto add a hard limit which is 10x (see HardLimitMultiplier)
+	// of soft limit.
+	//
+	// If the soft limit is violated,
+	// the violation will be reported via log.ErrorWithSentry,
+	// but it does not stop the normal parsing process.
+	//
+	// If the hard limit is violated,
+	// The loading of the file will fail immediately.
+	MaxFileSize int64 `yaml:"maxFileSize"`
+}
+
+// Get returns the latest parsed data from the file watcher.
+//
+// Although the type is interface{},
+// it's guaranteed to be whatever actual type is implemented inside Parser.
+func (r *Result) Get() interface{} {
+	return r.data.Load()
+}
+
+// Stop stops the file watcher.
+//
+// After Stop is called you won't get any updates on the file content,
+// but you can still call Get to get the last content before stopping.
+//
+// It's OK to call Stop multiple times.
+// Calls after the first one are essentially no-op.
+func (r *Result) Stop() {
+	r.cancel()
+}
+
+func (r *Result) watcherLoop(
+	watcher *fsnotify.Watcher,
+	parser filewatcher.Parser,
+	add AddFile,
+	remove RemoveFile,
+	softLimit, hardLimit int64,
+	logger log.Wrapper,
+) {
+	for {
+		select {
+		case <-r.ctx.Done():
+			watcher.Close()
+			return
+
+		case err := <-watcher.Errors:
+			logger.Log(context.Background(), "directorywatcher: watcher error: "+err.Error())
+
+		case ev := <-watcher.Events:
+
+			isDir, _ := isDirectory(ev.Name)
+
+			switch ev.Op {
+			default:
+				// do nothing.
+			case fsnotify.Create: // add to watcher, parse if file
+				// Wrap with an anonymous function to make sure that defer works.
+				func() {
+					if isDir {
+						watcher.Add(ev.Name)
+					} else {
+						f, err := limitopen.OpenWithLimit(ev.Name, softLimit, hardLimit)
+						if err != nil {
+							logger.Log(context.Background(), "directorywatcher: I/O error: "+err.Error())
```

Comment:
you are passing context.Background, which should only ever be used in main and test functions.  Always propagate your context.